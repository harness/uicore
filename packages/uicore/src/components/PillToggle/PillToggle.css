/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.optionBtns {
  display: inline-flex;
  align-items: center;
  border-radius: 100px;
  border: 1px solid var(--grey-400);
  font-size: var(--font-size-xsmall);
  font-weight: bold;
  height: 16px;
  box-sizing: border-box;

  .item {
    width: 62px;
    height: 16px;
    color: var(--grey-600);
    text-align: center;
    padding: 2px;
    border-radius: 100px;
    cursor: pointer;

    .disableToggleReasonIcon {
      margin: 0 var(--spacing-tiny);
    }
  }

  .selected {
    background-color: var(--grey-700);
    color: var(--white);
    cursor: default;
  }

  .disabledMode {
    cursor: default;
  }
}
