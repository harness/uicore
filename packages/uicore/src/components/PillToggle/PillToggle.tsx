/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import cx from 'classnames'
import { Classes, Intent, PopoverInteractionKind, Position } from '@blueprintjs/core'
import { Icon, IconName } from '@harness/icons'
import { Color } from '@harness/design-system'
import { Popover } from '../Popover/Popover'
import { Container } from '../Container/Container'
import { Layout } from '../../layouts/Layout'
import { Text } from '../Text/Text'
import css from './PillToggle.css'

export interface PillToggleOption<T> {
  label: string
  value: T
}

export interface PillToggleProps<T> {
  selectedView: T
  options: [PillToggleOption<T>, PillToggleOption<T>]
  onChange: (val: T) => void
  disableToggle?: boolean
  className?: string
  disableToggleReasonIcon?: IconName
  showDisableToggleReason?: boolean
  disableToggleReasonContent?: React.ReactElement
}

export const PillToggle = <T,>(props: PillToggleProps<T>): React.ReactElement => {
  const {
    selectedView,
    onChange,
    disableToggle = false,
    className = '',
    options,
    disableToggleReasonIcon = 'error-outline',
    showDisableToggleReason = false,
    disableToggleReasonContent
  } = props

  const renderInvalidBadge = (pillToggleValue: T) => {
    if (disableToggle && showDisableToggleReason && selectedView !== pillToggleValue) {
      return (
        <Popover interactionKind={PopoverInteractionKind.HOVER} position={Position.BOTTOM} className={Classes.DARK}>
          <Icon
            name={disableToggleReasonIcon}
            size={12}
            className={css.disableToggleReasonIcon}
            data-testid="invalid-icon"
            intent={Intent.DANGER}
          />
          {disableToggleReasonContent ?? (
            <Container padding="medium">
              <Layout.Vertical width={325} padding={{ left: 'small' }}>
                <Text
                  width={284}
                  color={Color.GREY_0}
                  margin={{ bottom: 'small' }}
                  font={{ size: 'normal', weight: 'light' }}>
                  The Visual Editor is disabled because of the errors in the YAML file.
                </Text>
                <Text width={284} color={Color.GREY_0} font={{ size: 'normal', weight: 'light' }}>
                  Fix all the errors indicated in the YAML Editor to enable the Visual mode.
                </Text>
              </Layout.Vertical>
            </Container>
          )}
        </Popover>
      )
    }
    return <></>
  }

  return (
    <div className={cx(css.optionBtns, className)}>
      <div
        data-name="toggle-option-one"
        className={cx(css.item, {
          [css.selected]: selectedView === options[0].value,
          [css.disabledMode]: disableToggle
        })}
        onClick={() => {
          if (selectedView === options[0].value || disableToggle) {
            return
          }
          onChange(options[0].value)
        }}
        tabIndex={0}
        role="button">
        {options[0].label}
        {renderInvalidBadge(options[0].value)}
      </div>
      <div
        data-name="toggle-option-two"
        className={cx(css.item, {
          [css.selected]: selectedView === options[1].value,
          [css.disabledMode]: disableToggle
        })}
        onClick={() => {
          if (selectedView === options[1].value || disableToggle) {
            return
          }
          onChange(options[1].value)
        }}
        tabIndex={0}
        role="button">
        {options[1].label}
        {renderInvalidBadge(options[1].value)}
      </div>
    </div>
  )
}
