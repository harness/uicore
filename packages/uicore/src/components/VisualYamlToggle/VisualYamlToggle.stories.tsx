/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { Meta, Story } from '@storybook/react'

import { VisualYamlToggle, VisualYamlToggleProps, VisualYamlSelectedView } from './VisualYamlToggle'

export default {
  title: 'Components / VisualYamlToggle',
  component: VisualYamlToggle
} as Meta

export const Basic: Story<VisualYamlToggleProps> = args => <VisualYamlToggle {...args} />
Basic.args = {
  selectedView: VisualYamlSelectedView.VISUAL
}
