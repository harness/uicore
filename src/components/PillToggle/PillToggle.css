.optionBtns {
  display: inline-flex;
  align-items: center;
  border-radius: 100px;
  border: 1px solid var(--grey-400);
  font-size: var(--font-size-xsmall);
  font-weight: bold;
  height: 16px;
  box-sizing: border-box;

  .item {
    width: 62px;
    height: 16px;
    color: var(--grey-600);
    text-align: center;
    padding: 2px;
    border-radius: 100px;
    cursor: pointer;
  }

  .selected {
    background-color: var(--grey-700);
    color: var(--white);
    cursor: default;
  }

  .disabledMode {
    pointer-events: none;
  }
}
