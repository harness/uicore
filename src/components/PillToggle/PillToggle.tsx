import React from 'react'
import cx from 'classnames'
import css from './PillToggle.css'

export interface PillToggleOption<T> {
  label: string
  value: T
}

export interface PillToggleProps<T> {
  selectedView?: T
  options: [PillToggleOption<T>, PillToggleOption<T>]
  onChange: (val: T) => void
  disableToggle?: boolean
  className?: string
}

export const PillToggle = <T,>(props: PillToggleProps<T>): React.ReactElement => {
  const { selectedView, onChange, disableToggle = false, className = '', options } = props
  const view = selectedView ?? options[0].value
  return (
    <div className={cx(css.optionBtns, className)}>
      <div
        data-name="toggle-option-one"
        className={cx(css.item, {
          [css.selected]: view === options[0].value,
          [css.disabledMode]: disableToggle
        })}
        onClick={() => {
          if (view === options[0].value) {
            return
          }
          onChange(options[0].value)
        }}
        tabIndex={0}
        role="button">
        {options[0].label}
      </div>
      <div
        data-name="toggle-option-two"
        className={cx(css.item, {
          [css.selected]: view === options[1].value,
          [css.disabledMode]: disableToggle
        })}
        onClick={() => {
          if (view === options[1].value) {
            return
          }
          onChange(options[1].value)
        }}
        tabIndex={0}
        role="button">
        {options[1].label}
      </div>
    </div>
  )
}
