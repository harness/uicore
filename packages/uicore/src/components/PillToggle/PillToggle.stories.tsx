/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { Meta, Story } from '@storybook/react'

import { PillToggle, PillToggleProps } from './PillToggle'

export default {
  title: 'Components / PillToggle',

  component: PillToggle
} as Meta

export const Basic: Story<PillToggleProps<string>> = args => <PillToggle {...args} />

Basic.args = {
  options: [
    { label: 'Label 1', value: 'label-1' },
    { label: 'Label 2', value: 'label-2' }
  ],
  selectedView: 'label-1'
}
