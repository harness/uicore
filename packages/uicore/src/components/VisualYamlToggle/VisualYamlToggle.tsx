/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { PillToggle, PillToggleProps } from '../PillToggle/PillToggle'

export enum VisualYamlSelectedView {
  VISUAL = 'VISUAL',
  YAML = 'YAML'
}

export interface VisualYamlToggleProps {
  selectedView: VisualYamlSelectedView
  onChange: (val: VisualYamlSelectedView) => void
  disableToggle?: boolean
  className?: string
  labels?: {
    visual?: string
    yaml?: string
  }
}

export function VisualYamlToggle(props: VisualYamlToggleProps): JSX.Element {
  const { selectedView, onChange, disableToggle = false, className = '', labels } = props

  const toggleProps: PillToggleProps<VisualYamlSelectedView> = {
    selectedView,
    options: [
      {
        label: labels?.visual || 'VISUAL',
        value: VisualYamlSelectedView.VISUAL
      },
      {
        label: labels?.yaml || 'YAML',
        value: VisualYamlSelectedView.YAML
      }
    ],
    onChange,
    disableToggle,
    className
  }

  return <PillToggle {...toggleProps} />
}
