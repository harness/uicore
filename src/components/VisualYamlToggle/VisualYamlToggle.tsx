import React from 'react'
import { PillToggle, PillToggleProps } from '../PillToggle/PillToggle'

export enum VisualYamlSelectedView {
  VISUAL = 'VISUAL',
  YAML = 'YAML'
}

export interface VisualYamlToggleProps {
  selectedView?: VisualYamlSelectedView
  onChange: (val: VisualYamlSelectedView) => void
  disableToggle?: boolean
  className?: string
  labels?: {
    visual?: string
    yaml?: string
  }
}

export function VisualYamlToggle(props: VisualYamlToggleProps): JSX.Element {
  const { selectedView, onChange, disableToggle = false, className = '', labels } = props

  const toggleProps: PillToggleProps<VisualYamlSelectedView> = {
    selectedView,
    options: [
      {
        label: labels?.visual || 'VISUAL',
        value: VisualYamlSelectedView.VISUAL
      },
      {
        label: labels?.yaml || 'YAML',
        value: VisualYamlSelectedView.YAML
      }
    ],
    onChange,
    disableToggle,
    className
  }

  return <PillToggle {...toggleProps} />
}
