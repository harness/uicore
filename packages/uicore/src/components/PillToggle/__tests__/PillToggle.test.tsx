/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, screen } from '@testing-library/react'
import { PillToggle } from '../PillToggle'
import { VisualYamlSelectedView } from 'components/VisualYamlToggle/VisualYamlToggle'
import userEvent from '@testing-library/user-event'

const options = [
  {
    label: 'VISUAL',
    value: VisualYamlSelectedView.VISUAL
  },
  {
    label: 'YAML',
    value: VisualYamlSelectedView.YAML
  }
]

describe('PillToggle', () => {
  test('snapshot testing on initial render', () => {
    const { container } = render(
      <PillToggle options={options as any} selectedView={VisualYamlSelectedView.VISUAL as any} onChange={jest.fn()} />
    )
    expect(container).toMatchSnapshot()
  })

  test('VISUAL should be selected', async () => {
    render(
      <PillToggle options={options as any} selectedView={VisualYamlSelectedView.VISUAL as any} onChange={jest.fn()} />
    )
    const visualDiv = screen.getByText('VISUAL')
    expect(visualDiv).toHaveClass('selected')
  })

  test('YAML should be selected', async () => {
    render(
      <PillToggle options={options as any} selectedView={VisualYamlSelectedView.YAML as any} onChange={jest.fn()} />
    )
    const yamlDiv = screen.getByText('YAML')
    expect(yamlDiv).toHaveClass('selected')
  })

  test('clicking on VISUAL should invoke onChange call', async () => {
    const onChange = jest.fn()
    render(
      <PillToggle options={options as any} selectedView={VisualYamlSelectedView.YAML as any} onChange={onChange} />
    )
    const visualDiv = screen.getByText('VISUAL')
    userEvent.click(visualDiv)
    expect(onChange).toHaveBeenCalled()
    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenCalledWith(VisualYamlSelectedView.VISUAL)
  })

  test('clicking on VISUAL should not invoke onChange call when toggle should be disabled', async () => {
    const onChange = jest.fn()
    render(
      <PillToggle
        options={options as any}
        selectedView={VisualYamlSelectedView.YAML as any}
        onChange={onChange}
        disableToggle={true}
      />
    )
    const visualDiv = screen.getByText('VISUAL')
    userEvent.click(visualDiv)
    expect(onChange).not.toHaveBeenCalled()
  })

  test('should show disable reason icon when showDisableToggleReason is true', async () => {
    const onChange = jest.fn()
    const { container } = render(
      <PillToggle
        options={options as any}
        selectedView={VisualYamlSelectedView.YAML as any}
        onChange={onChange}
        disableToggle={true}
        showDisableToggleReason={true}
      />
    )
    expect(container).toMatchSnapshot()
    const visualDiv = screen.getByText('VISUAL')
    userEvent.click(visualDiv)
    expect(onChange).not.toHaveBeenCalled()
  })
})
