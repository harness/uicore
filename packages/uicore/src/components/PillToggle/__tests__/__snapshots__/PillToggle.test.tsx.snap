// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PillToggle should show disable reason icon when showDisableToggleReason is true 1`] = `
<div>
  <div
    class="optionBtns"
  >
    <div
      class="item disabledMode"
      data-name="toggle-option-one"
      role="button"
      tabindex="0"
    >
      VISUAL
      <span
        class="bp3-popover-wrapper bp3-dark"
      >
        <span
          class="bp3-popover-target"
        >
          <span
            class="bp3-icon StyledProps--main disableToggleReasonIcon StyledProps--intent-danger"
            data-icon="error-outline"
            data-testid="invalid-icon"
            tabindex="0"
          >
            <svg
              fill="none"
              height="12"
              viewBox="0 0 16 16"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.423 1.631a.667.667 0 011.154 0l6.543 11.333a.667.667 0 01-.577 1H1.457a.667.667 0 01-.577-1L7.423 1.631z"
                fill="#fff"
              />
              <path
                clip-rule="evenodd"
                d="M7.57 2.253a.5.5 0 01.869 0l5.993 10.5a.5.5 0 01-.434.747H2.011a.5.5 0 01-.434-.748L7.57 2.253zm.407 8.242c.422 0 .763.333.763.744a.754.754 0 01-.763.745.754.754 0 01-.763-.745c0-.411.341-.745.763-.745zm.5-1.498a.5.5 0 11-1 0V5.498a.5.5 0 111 0v3.499z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
      </span>
    </div>
    <div
      class="item selected disabledMode"
      data-name="toggle-option-two"
      role="button"
      tabindex="0"
    >
      YAML
    </div>
  </div>
</div>
`;

exports[`PillToggle snapshot testing on initial render 1`] = `
<div>
  <div
    class="optionBtns"
  >
    <div
      class="item selected"
      data-name="toggle-option-one"
      role="button"
      tabindex="0"
    >
      VISUAL
    </div>
    <div
      class="item"
      data-name="toggle-option-two"
      role="button"
      tabindex="0"
    >
      YAML
    </div>
  </div>
</div>
`;
